contracts/contracts/chains/ethereum/EthereumMessengers.sol

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.12;
+
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/access/Ownable.sol";
+
+import "../../../interfaces/IMessenger.sol";
+import "../../RestrictedCalls.sol";
+import "../../Resolver.sol";
+
+contract EthereumL1Messenger is IMessenger, RestrictedCalls {
+    function callAllowed(address, address) external view returns (bool) {
+        // The call must be delivered by the Resolver
+        // Since we are on the same chain as the Resolver
+        // we should never see an execution of this function.
+        revert("Unexpected state.");
+    }
+
+    // Simply forward the message to the target on L1
+    /* solhint-disable avoid-tx-origin */
+    function sendMessage(address target, bytes calldata message)
+        external
+        restricted(block.chainid)
+    {
+        target.call(message);
+    }
+}
+
+// In order to keep naming consistent between our rollup specific contracts
+// we are using `EthereumL2Messenger` here as a name even if this contract
+// is supposed to be deployed on L1.
+contract EthereumL2Messenger is IMessenger, Ownable, RestrictedCalls {
+    enum MessageStatus {
+        UNDEFINED,
+        DEFINED,
+        RELAYED
+    }
+
+    Resolver public immutable resolver;
+
+    /// Maps message hashes with their status.
+    mapping(bytes32 => MessageStatus) public messageHashes;
+
+    constructor(address resolver_) {
+        resolver = Resolver(resolver_);
+    }
+
+    function callAllowed(address, address) external view returns (bool) {
+        // Since we are on the same chain as the FillManager & RequestManager
+        // we should never see an execution of this function.
+        revert("Unexpected state.");
+    }
+
+    // Stores the message hash for future use
+    function storeMessage(bytes calldata message) internal {
+        bytes32 messageHash = keccak256(message);
+
+        require(
+            messageHashes[messageHash] != MessageStatus.DEFINED,
+            "Message is already registered."
+        );
+        require(
+            messageHashes[messageHash] != MessageStatus.RELAYED,
+            "Message has already been relayed."
+        );
+
+        messageHashes[messageHash] = MessageStatus.DEFINED;
+    }
+
+    // Store message instead of forwarding it.
+    // Can be forwarded, if ever needed, by calling `relayMessage`.
+    //
+    // .. seealso:: :sol:func:`relayMessage`
+    function sendMessage(address, bytes calldata message)
+        external
+        restricted(block.chainid)
+    {
+        storeMessage(message);
```

Comment:
Is this inlined?

Reply:
> What does inlined mean here?

Does `solc`, the Solidity compiler, produce inlined code in this case, i.e. will it actually omit the `storeMessage` function and move its body into `sendMessage` in-line, as if it were written there?

Reply:
So, i tried running the solc compiler with `--ir-optimized` and i found out that the `storeMessage` is treated as another function that is being called inside `sendMessage`.

Should i move the body of the `storeMessage` inside of `sendMessage`? 
I only splitted this into its own function for better readability/testability of this contract.

I guess i can just squash the one into the other since we wont write any unit tests for this contract and instead include it in our E2E tests.

Reply:
I am currently assessing the difference in gas costs for the inlined an un-inlined version of this function to see if it really makes sense to inline the code.

Reply:
The result from the gas estimation made using `solc --gas` was `infinite gas` in both cases :D 
Any idea how i can get an absolute number here?

Reply:
> The result from the gas estimation made using `solc --gas` was `infinite gas` in both cases :D Any idea how i can get an absolute number here?

This is a know limitation in `solc` static gas analysis, please see e.g. https://ethereum.stackexchange.com/questions/29725/infinite-gas-estimation-from-solc-for-simple-function.